IOPool/Input/src/RepeatingCachedRootSource.cc

```diff
@@ -0,0 +1,368 @@
+
+  std::vector<std::shared_ptr<IndexIntoFile>> indexesIntoFiles(1);
+
+  auto input =
+      std::make_shared<InputFile>(physicalFileName.c_str(), "  Initiating request to open file ", InputType::Primary);
+  rootFile_ = makeRootFile(
+      logicalFileName, physicalFileName, 0 != nEventsToSkip, input, skipper, duplicateChecker, indexesIntoFiles);
+  rootFile_->reportOpened("repeating");
+
+  auto const& prodList = rootFile_->productRegistry()->productList();
+  productRegistryUpdate().updateFromInput(prodList);
+  //productRegistryUpdate().setFrozen();
+
+  //setup caching
+  auto nProdsInEvent =
+      std::count_if(prodList.begin(), prodList.end(), [](auto&& iV) { return iV.second.branchType() == edm::InEvent; });
+  {
+    size_t index = 0;
+    for (auto& prod : prodList) {
+      if (prod.second.branchType() == edm::InEvent) {
+        branchIDToWrapperIndex_[prod.second.branchID()] = index++;
+      }
+    }
+  }
+  for (auto& cache : cachedWrappers_) {
+    cache.resize(nProdsInEvent);
+  }
+}
+
+void RepeatingCachedRootSource::beginJob() {
+  ProcessConfiguration processConfiguration;
+  processConfiguration.setParameterSetID(ParameterSet::emptyParameterSetID());
+  processConfiguration.setProcessConfigurationID();
+
+  EventPrincipal eventPrincipal(productRegistry(),
+                                std::make_shared<BranchIDListHelper>(),
+                                std::make_shared<ThinnedAssociationsHelper>(),
+                                processConfiguration,
+                                nullptr);
+
+  {
+    RunNumber_t run = 0;
+    LuminosityBlockNumber_t lumi = 0;
+    auto itAux = eventAuxs_.begin();
+    for (auto& cache : cachedWrappers_) {
+      rootFile_->nextEventEntry();
+      rootFile_->readCurrentEvent(eventPrincipal);
+      auto const& aux = eventPrincipal.aux();
+      *(itAux++) = aux;
+      if (0 == run) {
+        run = aux.run();
+        lumi = aux.luminosityBlock();
+      } else {
+        if (run != aux.run()) {
+          throw cms::Exception("EventsWithDifferentRuns") << "The requested events to cache are from different Runs";
+        }
+        if (lumi != aux.luminosityBlock()) {
+          throw cms::Exception("EventsWithDifferentLuminosityBlocks")
+              << "The requested events to cache are from different LuminosityBlocks";
+        }
+      }
+      selectionIDs_ = eventPrincipal.eventSelectionIDs();
+      branchListIndexes_ = eventPrincipal.branchListIndexes();
+      {
+        auto reader = eventPrincipal.reader();
+        for (auto const& branchToIndex : branchIDToWrapperIndex_) {
+          cache[branchToIndex.second] = reader->getProduct(branchToIndex.first, &eventPrincipal);
+        }
+      }
+    }
+    rootFile_->rewind();
+  }
+}
+
+void RepeatingCachedRootSource::fillDescriptions(ConfigurationDescriptions& descriptions) {
+  ParameterSetDescription desc;
+  desc.setComment("Read only a few Events from EDM/Root files.");
```

Comment:
I'd suggest to expand with something along
```suggestion
  desc.setComment("Read only a few Events from one EDM/Root file, and repeat them in sequence. The Events are required to be from the same Run and LuminosityBlock.");
```

Reply:
done